### PR TITLE
ingester: Support for disallowing Push API for ingest storage

### PR DESCRIPTION
**Pull request by narqo**
#### What this PR does

~~_These changes are still in progress by I wanted to get some early feedback, on if the direction is right_~~

The changes add a new flag `-ingerster.push-grpc-method-enabled (default "true")`. It allows a user to configure the app to block "push" gRPC API in the ingester module, IFF it has ingest-store enabled.

#### Which issue(s) this PR fixes or relates to

n/a

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/ingester/ingester.go [hunk 1]
```diff
@@ -231,6 +234,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
 	f.BoolVar(&cfg.UpdateIngesterOwnedSeries, "ingester.track-ingester-owned-series", false, "This option enables tracking of ingester-owned series based on ring state, even if -ingester.use-ingester-owned-series-for-limits is disabled.")
 	f.DurationVar(&cfg.OwnedSeriesUpdateInterval, "ingester.owned-series-update-interval", 15*time.Second, "How often to check for ring changes and possibly recompute owned series as a result of detected change.")
 
+	f.BoolVar(&cfg.EnablePushAPI, "ingester.enable-push-api", true, "When disabled, disallows Push API in ingester. Use after migrating ingester to ingest-storage.")
```
Comment (narqo) [inline pkg/ingester/ingester.go:237]: I'm not too fancy about the phrasing here, and am open for better options, that would sound `help`-ful :)

File: Makefile [hunk 2]
```diff
@@ -615,10 +615,10 @@ check-doc-validator: ## Check documentation using doc-validator tool
 
 .PHONY: reference-help
 reference-help: ## Generates the reference help documentation.
-reference-help: cmd/mimir/mimir
+reference-help: cmd/mimir/mimir tools/config-inspector/config-inspector
 	@(./cmd/mimir/mimir -h || true) > cmd/mimir/help.txt.tmpl
 	@(./cmd/mimir/mimir -help-all || true) > cmd/mimir/help-all.txt.tmpl
-	@(go run ./tools/config-inspector || true) > cmd/mimir/config-descriptor.json
+	@(./tools/config-inspector/config-inspector || true) > cmd/mimir/config-descriptor.json
```
Comment (narqo) [inline Makefile:621]: This one doesn't directly belong to the changes. It's here so the re-generated config on my local machine (that uses Go 1.22), matched the one on the CI (that still runs Go 1.21).

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -206,6 +206,9 @@ type Config struct {
 	UpdateIngesterOwnedSeries       bool          `yaml:"track_ingester_owned_series" category:"experimental"`
 	OwnedSeriesUpdateInterval       time.Duration `yaml:"owned_series_update_interval" category:"experimental"`
 
+	// TODO Make "false" by default, after IngestStorageConfig.Enabled is stable
```
Comment (pracucci) [inline pkg/ingester/ingester.go:209]: Please remove this comment for two reasons:
1. We can safely switch to false only once ingest storage will be the default (but it's very far from now)
2. We've a bunch of other config that will be changed once ingest storage will be the default, so we can address all of them later (in a far future...)

File: pkg/ingester/ingester.go [hunk 4]
```diff
@@ -206,6 +206,9 @@ type Config struct {
...
+	EnablePushAPI bool `yaml:"enable_push_api" category:"experimental"`
```
Comment (pracucci) [inline pkg/ingester/ingester.go:210]: So far we're keeping hidden all the ingest storage config.

```suggestion
	EnablePushAPI bool `yaml:"enable_push_api" category:"experimental" doc:"hidden"`
```

File: pkg/ingester/errors.go [hunk 5]
```diff
@@ -525,6 +527,19 @@ func (e ingesterTooBusyError) errorCause() mimirpb.ErrorCause {
...
 var _ ingesterError = ingesterTooBusyError{}
 
+type ingesterPushAPIDisabledError struct{}
+
+func (e ingesterPushAPIDisabledError) Error() string {
+	return ingesterPushAPIDisabledMsg
+}
+
+func (e ingesterPushAPIDisabledError) errorCause() mimirpb.ErrorCause {
+	return mimirpb.SERVICE_UNAVAILABLE
```
Comment (pracucci) [inline pkg/ingester/errors.go:537]: `SERVICE_UNAVAILABLE` is treated as a server-side error, so retryable. I'm wondering if we should add a `NOT_ALLOWED` error, and then we map it as `405` status code and treat it as non retryable.

File: pkg/ingester/ingester.go [hunk 6]
```diff
@@ -954,6 +963,10 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReques
 	// retain anything from `req` past the exit from this function.
 	defer cleanUp()
 
+	if !i.cfg.EnablePushAPI {
```
Comment (pracucci) [inline pkg/ingester/ingester.go:966]: We need to allow `PushWithCleanup()` because that's what we call to ingest data from Kafka (see `NewPartitionReaderForPusher()` called from `ingester.New()`).

Why not existing test on ingest storage failed (even if we have very few, at least one should have failed). Reason is that we're not disabling the Push API in ingest storage tests. Please set `EnablePushAPI` to `false` in `createTestIngesterWithIngestStorage()`. You should get at least `TestIngester_QueryStream_IngestStorageReadConsistency` failing.

So how to solve it? We could be tempted to remove this check and keep `StartPushRequest()`, but `StartPushRequest()` is called if `LimitInflightRequestsUsingGrpcMethodLimiter` is disabled (and BTW we need to keep `LimitInflightRequestsUsingGrpcMethodLimiter` disabled with ingest storage because there's no ingestion via gRPC).

I think to solve it easily (hopefully) we have two move the check out of `StartPushRequest()` and `PushWithCleanup()` into ` Push()`, move the current `Push()` content into `push()` and then call `push()` when we ingest data via Kafka (or something similar, you got the idea).

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! I left a major comment.

File: pkg/ingester/ingester.go (hunk 6 above)
Comment (narqo) [reply pkg/ingester/ingester.go]: I decided to try re-working this differently: instead of disabling the functionality from the inside of `Ingester`, I do that in `pkg/mimir` (see the changes below), closer to where we register an implementation of the `api.Ingester` (the actual gRPC handlers) in the gRPC server.

I feel this will be less prone to human errors. What do you think? 

Note I haven't updated the tests, yet; just wanted to get feedback, first

cc @pstibrany

Comment (pstibrany) [reply pkg/ingester/ingester.go]: Wrapping ingester to disable its `Push` gRPC method sounds good to me. This way the test changes that Marco asked for in his comment will not be easy to do, but I don't think that's a blocker.

Let's please move that `apiIngester` type into `pkg/ingester` (I'd also recommend finding a better name for it, eg. `ingesterWithDisabledPush`?), and minimize unrelated changes in this PR.

File: pkg/ingester/ingester.go (hunk 1 above)
Comment (pstibrany) [reply pkg/ingester/ingester.go:236]: Maybe something like this?

```suggestion
	f.BoolVar(&cfg.EnablePushAPI, "ingester.enable-push-api", true, "Enables Push gRPC method on ingester. This can only be disabled when using ingest-storage to make sure that ingesters only receive data from Kafka.")
```

File: pkg/ingester/ingester.go [hunk 7]
```diff
@@ -3487,14 +3491,26 @@ func (i *Ingester) checkAvailable() error {
 	return newUnavailableError(s)
 }
 
+// PushToStorage is a low-level implementation of Push() for block storage.
```
Comment (pstibrany) [inline pkg/ingester/ingester.go:3494]:
```suggestion
// PushToStorage implements ingest.Pusher interface for ingestion via ingest-storage.
```

File: pkg/ingester/ingester.go [hunk 8]
```diff
@@ -3487,14 +3491,26 @@ func (i *Ingester) checkAvailable() error {
...
+func (i *Ingester) PushToStorage(ctx context.Context, req *mimirpb.WriteRequest) error {
+	err := i.PushWithCleanup(ctx, req, func() { mimirpb.ReuseSlice(req.Timeseries) })
+	if err != nil {
+		return i.mapPushErrorToErrorWithStatus(err)
+	}
+	return nil
+}
+
 // Push implements client.IngesterServer
```
Comment (pstibrany) [inline pkg/ingester/ingester.go:3503]: We could update this comment to mention that this gRPC method. Maybe something like

```suggestion
// Push implements client.IngesterServer, which is registered into gRPC server. 
```

File: pkg/mimir/grpc_push_check.go [hunk 9]
```diff
@@ -68,6 +68,9 @@ func (g *grpcInflightMethodLimiter) RPCCallStarting(ctx context.Context, methodN
 
 		err := ing.StartPushRequest(reqSize)
 		if err != nil {
+			if st, ok := grpcutil.ErrorToStatus(err); ok {
+				return ctx, st.Err()
+			}
```
Comment (pstibrany) [inline pkg/mimir/grpc_push_check.go:73]: Let's please revert this unnecessary change.

File: pkg/mimir/mimir.go [hunk 10]
```diff
@@ -240,8 +240,13 @@ func (c *Config) Validate(log log.Logger) error {
...
 		return errors.Wrap(err, "invalid ingest storage config")
 	}
-	if c.isAnyModuleEnabled(Ingester, Write, All) && c.IngestStorage.Enabled && !c.Ingester.DeprecatedReturnOnlyGRPCErrors {
-		return errors.New("to use ingest storage (-ingest-storage.enabled) also enable -ingester.return-only-grpc-errors")
+	if c.isAnyModuleEnabled(Ingester, Write, All) {
+		if c.IngestStorage.Enabled && !c.Ingester.DeprecatedReturnOnlyGRPCErrors {
+			return errors.New("to use ingest storage (-ingest-storage.enabled) also enable -ingester.return-only-grpc-errors")
+		}
+		if !c.IngestStorage.Enabled && !c.Ingester.EnablePushAPI {
+			return errors.New("cannot disable Push API in ingester, while ingest storage (-ingest-storage.enabled) is also disabled")
```
Comment (pstibrany) [inline pkg/mimir/mimir.go:248]: suggestion: "cannot disable Push gRPC method in ingester, because ingest storage is not enabled"

**Review by pstibrany — COMMENTED** [review]
Thank you, this looks good overall. I have left some small comments.

I would not call this "Push API" because that is widely understood as HTTP push endpoint on distributors. Ingesters only have "Push" gRPC method, so I would suggest to use that name explicitely.

I would also like to see changes that are not necessary for this PR in ingester error mapping and grpc push check to be reverted.

**Comment by narqo** [comment]
🆙

I think, I've addressed all the comments. PHAL

**Review by pstibrany — APPROVED** [review]
Thank you very much for this PR and addressing my feedback.

File: pkg/mimirpb/mimir.proto [hunk 11]
```diff
@@ -46,6 +46,7 @@ enum ErrorCause {
   TSDB_UNAVAILABLE = 8;
   TOO_BUSY = 9;
   CIRCUIT_BREAKER_OPEN = 10;
+  METHOD_NOT_ALLOWED = 11;
```
Comment (pracucci) [inline pkg/mimirpb/mimir.proto:49]: We've added a new error cause, but then this has to be explicitly mapped to HTTP status codes. As a reference, look all places where `TSDB_UNAVAILABLE` is used and do something similar for `METHOD_NOT_ALLOWED` too. I would return `405` for `METHOD_NOT_ALLOWED`.

Comment (pracucci) [reply pkg/mimirpb/mimir.proto:49]: Actually I see you proposed it in https://github.com/grafana/mimir/pull/7618 but got rejected with a good feedback. Let me follow up offline about it.
